File: src/daytrace/transform.py
from __future__ import annotations

from collections import defaultdict
from collections.abc import Iterable
from dataclasses import replace
from datetime import datetime, timedelta

from daytrace.models import ActivityRecord, SourceKind


def _sort(records: Iterable[ActivityRecord]) -> tuple[ActivityRecord, ...]:
    return tuple(
        sorted(records, key=lambda item: (item.start, item.bucket_id, item.event_id))
    )


def _union(
    intervals: Iterable[tuple[datetime, datetime]],
) -> list[tuple[datetime, datetime]]:
    merged: list[tuple[datetime, datetime]] = []
    for start, end in sorted(intervals):
        if not merged or start > merged[-1][1]:
            merged.append((start, end))
        else:
            merged[-1] = (merged[-1][0], max(merged[-1][1], end))
    return merged


def _subtract_interval(
    fragments: list[tuple[datetime, datetime]],
    away_start: datetime,
    away_end: datetime,
) -> list[tuple[datetime, datetime]]:
    output: list[tuple[datetime, datetime]] = []
    for start, end in fragments:
        if away_end <= start or away_start >= end:
            output.append((start, end))
            continue
        if start < away_start:
            output.append((start, away_start))
        if away_end < end:
            output.append((away_end, end))
    return output


def remove_afk(records: Iterable[ActivityRecord]) -> tuple[ActivityRecord, ...]:
    items = tuple(records)
    away = _union(
        (item.start, item.end)
        for item in items
        if item.kind is SourceKind.AFK and (item.status or "").casefold() == "afk"
    )
    output: list[ActivityRecord] = []
    for item in items:
        if item.kind is SourceKind.AFK:
            continue
        fragments = [(item.start, item.end)]
        for away_start, away_end in away:
            fragments = _subtract_interval(fragments, away_start, away_end)
        output.extend(replace(item, start=start, end=end) for start, end in fragments)
    return _sort(output)


def filter_project(
    records: Iterable[ActivityRecord], project: str | None
) -> tuple[ActivityRecord, ...]:
    if project is None:
        return _sort(records)
    needle = project.casefold()
    return _sort(
        item
        for item in records
        if any(
            needle in value.casefold()
            for value in (
                item.project,
                item.app,
                item.title,
                item.file,
                item.url_host,
                item.bucket_id,
            )
            if value
        )
    )


def merge_adjacent(
    records: Iterable[ActivityRecord],
    maximum_gap: timedelta = timedelta(seconds=60),
) -> tuple[ActivityRecord, ...]:
    by_bucket: dict[str, list[ActivityRecord]] = defaultdict(list)
    for item in _sort(records):
        bucket_records = by_bucket[item.bucket_id]
        if (
            bucket_records
            and bucket_records[-1].content_key() == item.content_key()
            and item.start - bucket_records[-1].end <= maximum_gap
        ):
            bucket_records[-1] = replace(
                bucket_records[-1], end=max(bucket_records[-1].end, item.end)
            )
        else:
            bucket_records.append(item)
    return _sort(
        item for bucket_records in by_bucket.values() for item in bucket_records
    )


def partition_window_seconds(
    records: Iterable[ActivityRecord],
) -> tuple[tuple[ActivityRecord, float], ...]:
    windows = tuple(item for item in records if item.kind is SourceKind.WINDOW)
    boundaries = sorted({point for item in windows for point in (item.start, item.end)})
    totals: dict[ActivityRecord, float] = defaultdict(float)
    for start, end in zip(boundaries, boundaries[1:]):
        candidates = [item for item in windows if item.start < end and item.end > start]
        if candidates:
            winner = min(candidates, key=lambda item: (item.bucket_id, item.event_id))
            totals[winner] += (end - start).total_seconds()
    return tuple(
        sorted(totals.items(), key=lambda pair: (pair[0].bucket_id, pair[0].event_id))
    )

File: packages/daytrace-core/src/transform.ts
import type { ActivityRecord } from "./models.js";
import { timestampMs } from "./time.js";

type Interval = readonly [number, number];

function compareRecords(left: ActivityRecord, right: ActivityRecord): number {
  return timestampMs(left.start) - timestampMs(right.start)
    || left.bucketId.localeCompare(right.bucketId)
    || left.eventId.localeCompare(right.eventId);
}

function sorted(records: readonly ActivityRecord[]): ActivityRecord[] {
  return [...records].sort(compareRecords);
}

function union(intervals: readonly Interval[]): Interval[] {
  const merged: Interval[] = [];
  for (const [start, end] of [...intervals].sort((left, right) => left[0] - right[0] || left[1] - right[1])) {
    const previous = merged.at(-1);
    if (previous === undefined || start > previous[1]) {
      merged.push([start, end]);
    } else {
      merged[merged.length - 1] = [previous[0], Math.max(previous[1], end)];
    }
  }
  return merged;
}

function subtractInterval(fragments: readonly Interval[], awayStart: number, awayEnd: number): Interval[] {
  const output: Interval[] = [];
  for (const [start, end] of fragments) {
    if (awayEnd <= start || awayStart >= end) {
      output.push([start, end]);
      continue;
    }
    if (start < awayStart) output.push([start, awayStart]);
    if (awayEnd < end) output.push([awayEnd, end]);
  }
  return output;
}

export function removeAfk(records: readonly ActivityRecord[]): readonly ActivityRecord[] {
  const away = union(records
    .filter((item) => item.kind === "afk" && (item.status ?? "").toLocaleLowerCase() === "afk")
    .map((item) => [timestampMs(item.start), timestampMs(item.end)] as const));
  const output: ActivityRecord[] = [];
  for (const item of records) {
    if (item.kind === "afk") continue;
    let fragments: Interval[] = [[timestampMs(item.start), timestampMs(item.end)]];
    for (const [awayStart, awayEnd] of away) {
      fragments = subtractInterval(fragments, awayStart, awayEnd);
    }
    for (const [start, end] of fragments) {
      output.push({ ...item, start: new Date(start).toISOString(), end: new Date(end).toISOString() });
    }
  }
  return output.sort(compareRecords);
}

export function filterProject(records: readonly ActivityRecord[], project?: string): readonly ActivityRecord[] {
  if (project === undefined) return sorted(records);
  const needle = project.toLocaleLowerCase();
  return sorted(records.filter((item) => [
    item.project,
    item.app,
    item.title,
    item.file,
    item.urlHost,
    item.bucketId,
  ].some((value) => value?.toLocaleLowerCase().includes(needle) === true)));
}

function contentKey(item: ActivityRecord): string {
  return JSON.stringify([
    item.bucketId,
    item.kind,
    item.app ?? null,
    item.title ?? null,
    item.project ?? null,
    item.file ?? null,
    item.urlHost ?? null,
    item.urlPath ?? null,
    item.language ?? null,
    item.status ?? null,
  ]);
}

export function mergeAdjacent(
  records: readonly ActivityRecord[],
  maximumGapSeconds = 60,
): readonly ActivityRecord[] {
  const byBucket = new Map<string, ActivityRecord[]>();
  for (const item of sorted(records)) {
    const bucket = byBucket.get(item.bucketId) ?? [];
    const previous = bucket.at(-1);
    if (
      previous !== undefined
      && contentKey(previous) === contentKey(item)
      && timestampMs(item.start) - timestampMs(previous.end) <= maximumGapSeconds * 1000
    ) {
      bucket[bucket.length - 1] = {
        ...previous,
        end: timestampMs(previous.end) >= timestampMs(item.end) ? previous.end : item.end,
      };
    } else {
      bucket.push(item);
    }
    byBucket.set(item.bucketId, bucket);
  }
  return [...byBucket.values()].flat().sort(compareRecords);
}

export interface WindowAllocation {
  readonly record: ActivityRecord;
  readonly seconds: number;
}

export function partitionWindowSeconds(records: readonly ActivityRecord[]): readonly WindowAllocation[] {
  const windows = records.filter((item) => item.kind === "current-window");
  const boundaries = [...new Set(windows.flatMap((item) => [timestampMs(item.start), timestampMs(item.end)]))]
    .sort((left, right) => left - right);
  const totals = new Map<ActivityRecord, number>();
  for (let index = 0; index < boundaries.length - 1; index += 1) {
    const start = boundaries[index];
    const end = boundaries[index + 1];
    if (start === undefined || end === undefined) continue;
    const candidates = windows.filter((item) => timestampMs(item.start) < end && timestampMs(item.end) > start)
      .sort((left, right) => left.bucketId.localeCompare(right.bucketId) || left.eventId.localeCompare(right.eventId));
    const winner = candidates[0];
    if (winner !== undefined) totals.set(winner, (totals.get(winner) ?? 0) + (end - start) / 1000);
  }
  return [...totals.entries()]
    .sort(([left], [right]) => left.bucketId.localeCompare(right.bucketId) || left.eventId.localeCompare(right.eventId))
    .map(([record, seconds]) => ({ record, seconds }));
}
